### ISSUE TITLE
Rule Change: [no-loop-func] don’t warn when control flow will always leave the loop

### ISSUE THREAD

**Issue by j-f1** [eslint/eslint#16954]
### What rule do you want to change?

no-loop-func

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```js
function runComputations(idx = 0) {
  while (idx < this.computeQueue.length) {
    let result = compute(idx);
    if (result instanceof Promise) {
      result.then(value => {
        runComputations(idx + 1);
      });
      return;
    }
    idx++;
  }
}
```

### What does the rule currently do for this code?

Warns on the callback passed to `then`

### What will the rule do after it's changed?

Identify that there is no way for the value of `idx` to aft the callback is created (since control flow moves out of the function, and the only thing that changes `idx` is outside of that path).

### Participation

- [X] I am willing to submit a pull request to implement this change.

**Comment by nzakas** [comment]
We discussed in the 2023-03-23 TSC meeting and even though this is a false positive, the complexity required to catch this edge case isn't worth the extra effort or performance cost, so closing. [Meeting Notes](https://github.com/eslint/tsc-meetings/pull/419)
